File: backend/python/app/services/implementations/entity_service_mg.py
from ..interfaces.file_storage_service import IFileStorageService
from ...models.entity_mg import Entity
from ..interfaces.entity_service import IEntityService
from uuid import uuid4


class EntityService(IEntityService):
    def __init__(self, logger, file_storage_service: IFileStorageService):
        self.logger = logger
        self.file_storage_service = file_storage_service

    def get_entities(self):
        entities = []

        for result in Entity.objects:
            result_dict = result.to_serializable_dict()
            entities.append(result_dict)

        return entities

    def get_entity(self, id):
        entity = Entity.objects.get(id=id)

        if entity is None:
            self.logger.error("Invalid id")
            raise Exception("Invalid id")

        return entity.to_serializable_dict()

    def create_entity(self, entity):
        try:
            file_name = None
            if entity.file:
                file_name = str(uuid4())
                self.file_storage_service.create_file(
                    file_name, entity.file, entity.file.content_type
                )
            entity.__dict__.pop("file", None)
            new_entity = Entity(**entity.__dict__, file_name=file_name)
            new_entity.save()
        except Exception as error:
            self.logger.error(str(error))
            raise error

        return new_entity.to_serializable_dict()

    def update_entity(self, id, entity):
        current_entity = Entity.objects.get(id=id)
        if current_entity is None:
            self.logger.error("Invalid id")
            raise Exception("Invalid id")
        file_name = current_entity.file_name
        if entity.file:
            if file_name:
                self.file_storage_service.update_file(
                    file_name, entity.file, entity.file.content_type
                )
            else:
                file_name = str(uuid4())
                self.file_storage_service.create_file(
                    file_name, entity.file, entity.file.content_type
                )
        elif file_name:
            self.file_storage_service.delete_file(file_name)
            file_name = None

        entity.__dict__.pop("file", None)
        updated_entity = Entity.objects(id=id).modify(
            new=True, **entity.__dict__, file_name=file_name
        )

        if updated_entity is None:
            self.logger.error("Invalid id")
            raise Exception("Invalid id")

        return updated_entity.to_serializable_dict()

    def delete_entity(self, id):
        try:
            file_name = Entity.objects.get(id=id).file_name
            deleted = Entity.objects(id=id).modify(remove=True)
            if deleted:
                self.file_storage_service.delete_file(file_name)
            return id
        except Exception as error:
            self.logger.error(str(error))
            raise error

File: backend/typescript/services/implementations/EntityServiceMg.ts
import { v4 as uuidv4 } from "uuid";

import MgEntity, { Entity } from "../../models/entity.mgmodel";
import {
  IEntityService,
  EntityRequestDTO,
  EntityResponseDTO,
} from "../interfaces/IEntityService";
import IFileStorageService from "../interfaces/storageService";
import logger from "../../utilities/logger";

const Logger = logger(__filename);

class EntityService implements IEntityService {
  storageService: IFileStorageService;

  constructor(storageService: IFileStorageService) {
    this.storageService = storageService;
  }

  /* eslint-disable class-methods-use-this */
  async getEntity(id: string): Promise<EntityResponseDTO> {
    let entity: Entity | null;
    try {
      entity = await MgEntity.findById(id);
      if (!entity) {
        throw new Error(`Entity id ${id} not found`);
      }
    } catch (error) {
      Logger.error(`Failed to get entity. Reason = ${error.message}`);
      throw error;
    }

    return {
      id: entity.id,
      stringField: entity.stringField,
      intField: entity.intField,
      enumField: entity.enumField,
      stringArrayField: entity.stringArrayField,
      boolField: entity.boolField,
      fileName: entity.fileName,
    };
  }

  async getEntities(): Promise<EntityResponseDTO[]> {
    try {
      const entities: Array<Entity> = await MgEntity.find();
      return entities.map((entity) => ({
        id: entity.id,
        stringField: entity.stringField,
        intField: entity.intField,
        enumField: entity.enumField,
        stringArrayField: entity.stringArrayField,
        boolField: entity.boolField,
        fileName: entity.fileName,
      }));
    } catch (error) {
      Logger.error(`Failed to get entities. Reason = ${error.message}`);
      throw error;
    }
  }

  async createEntity(entity: EntityRequestDTO): Promise<EntityResponseDTO> {
    let newEntity: Entity | null;
    const fileName = entity.filePath ? uuidv4() : "";
    try {
      newEntity = await MgEntity.create({ ...entity, fileName });
      if (entity.filePath) {
        this.storageService.createFile(
          fileName,
          entity.filePath,
          entity.fileContentType,
        );
      }
    } catch (error) {
      Logger.error(`Failed to create entity. Reason = ${error.message}`);
      throw error;
    }
    return {
      id: newEntity.id,
      stringField: newEntity.stringField,
      intField: newEntity.intField,
      enumField: newEntity.enumField,
      stringArrayField: newEntity.stringArrayField,
      boolField: newEntity.boolField,
      fileName,
    };
  }

  async updateEntity(
    id: string,
    entity: EntityRequestDTO,
  ): Promise<EntityResponseDTO | null> {
    let updatedEntity: Entity | null;
    let fileName = "";
    try {
      const currentEntity = await MgEntity.findById(id, "fileName");
      const currentFileName = currentEntity?.fileName;
      if (entity.filePath) {
        fileName = currentFileName || uuidv4();
      }
      updatedEntity = await MgEntity.findByIdAndUpdate(
        id,
        { ...entity, fileName },
        {
          new: true,
          runValidators: true,
        },
      );
      if (!updatedEntity) {
        throw new Error(`Entity id ${id} not found`);
      }
      if (entity.filePath) {
        if (currentFileName) {
          this.storageService.updateFile(
            fileName,
            entity.filePath,
            entity.fileContentType,
          );
        } else {
          this.storageService.createFile(
            fileName,
            entity.filePath,
            entity.fileContentType,
          );
        }
      } else if (currentFileName) {
        this.storageService.deleteFile(currentFileName);
      }
    } catch (error) {
      Logger.error(`Failed to update entity. Reason = ${error.message}`);
      throw error;
    }
    return {
      id: updatedEntity.id,
      stringField: updatedEntity.stringField,
      intField: updatedEntity.intField,
      enumField: updatedEntity.enumField,
      stringArrayField: updatedEntity.stringArrayField,
      boolField: updatedEntity.boolField,
      fileName,
    };
  }

  async deleteEntity(id: string): Promise<void> {
    try {
      const deletedEntity: Entity | null = await MgEntity.findByIdAndDelete(id);
      if (!deletedEntity) {
        throw new Error(`Entity id ${id} not found`);
      }
      this.storageService.deleteFile(deletedEntity.fileName);
    } catch (error) {
      Logger.error(`Failed to delete entity. Reason = ${error.message}`);
      throw error;
    }
  }
}

export default EntityService;
